htdocs/core/class/commonsignedobject.php

```diff
@@ -0,0 +1,203 @@
+		$langs->load("commercial");
+
+		$l10n_signed_status_labels = [
+			self::$SIGNED_STATUSES['STATUS_NO_SIGNATURE']			=> 'NoSignature',
+			self::$SIGNED_STATUSES['STATUS_SIGNED_SENDER']			=> 'SignedSender',
+			self::$SIGNED_STATUSES['STATUS_SIGNED_RECEIVER']		=> 'SignedReceiver',
+			self::$SIGNED_STATUSES['STATUS_SIGNED_RECEIVER_ONLINE']	=> 'SignedReceiverOnline',
+			self::$SIGNED_STATUSES['STATUS_SIGNED_ALL']				=> 'SignedAll'
+		];
+
+		$l10n_signed_status = [];
+		foreach (self::$SIGNED_STATUSES as $signed_status_code) {
+			$l10n_signed_status[$signed_status_code] = $langs->transnoentitiesnoconv($l10n_signed_status_labels[$signed_status_code]);
+		}
+		return $l10n_signed_status;
+	}
+
+	/**
+	 * Set signed status & object context. Call sign action trigger.
+	 *
+	 * @param	User	$user			Object user that modify
+	 * @param	int		$status			New signed status to set (often a constant like self::STATUS_XXX)
+	 * @param	int		$notrigger		1 = Does not execute triggers, 0 = Execute triggers
+	 * @param	string	$triggercode	Trigger code to use
+	 * @return	int						0 < if KO, > 0 if OK
+	 */
+	public function setSignedStatus(User $user, int $status = 0, int $notrigger = 0, string $triggercode = ''): int
+	{
+		global $langs;
+		$langs->loadLangs(array('commercial'));
+		$this->signed_status = $status;
+		$this->context['signature'] = $status;
+		switch ($status) {
+			case 0:
+				$this->context['actionmsg2'] = $langs->transnoentitiesnoconv('UnsignedInDolibarr');
+				break;
+			case 1:
+				$this->context['actionmsg2'] = $langs->transnoentitiesnoconv('SignedSender');
+				break;
+			case 2:
+				$this->context['actionmsg2'] = $langs->transnoentitiesnoconv('SignedReceiver');
+				break;
+			case 3:
+				$this->context['actionmsg2'] = $langs->transnoentitiesnoconv('SignedReceiverOnline');
+				break;
+			case 9:
+				$this->context['actionmsg2'] = $langs->transnoentitiesnoconv('SignedAll');
+				break;
+		}
+		return $this->setSignedStatusCommon($user, $status, $notrigger, $triggercode);
+	}
+
+	/**
+	 * Set signed status & call trigger with context message
+	 *
+	 * @param	User	$user			Object user that modify
+	 * @param	int		$status			New signed status to set (often a constant like self::STATUS_XXX)
+	 * @param	int		$notrigger		1 = Does not execute triggers, 0 = Execute triggers
+	 * @param	string	$triggercode	Trigger code to use
+	 * @return	int						0 < if KO, > 0 if OK
+	 */
+	public function setSignedStatusCommon(User $user, int $status, int $notrigger = 0, string $triggercode = ''): int
+	{
+		$error = 0;
+
+		$this->db->begin();
+
+		$statusfield = 'signed_status';
+
+		$sql = "UPDATE ".$this->db->prefix().$this->table_element;
+		$sql .= " SET ".$statusfield." = ".((int) $status);
+		$sql .= " WHERE rowid = ".((int) $this->id);
+
+		if ($this->db->query($sql)) {
+			if (!$error) {
+				$this->oldcopy = clone $this;
+			}
+
+			if (!$error && !$notrigger) {
+				// Call trigger
+				$result = $this->call_trigger($triggercode, $user);
```

Comment:
This supposes that the Trait is used with a CommonObject and the Trait can not itself be restricted to certain classes language-wise.

There are two solutions:
- Test that $this is a CommonObject (`if ($this instanceof CommonObject`) and preferable add some error handling;
- Add phpdoc suggesting that the method exists (which can also serve as documentation to remind that the method is expected on class the trait is used with).
```
Methods that should exist in the host class:
@method call_trigger() <returntype> call_trigger(<param doc>)

Reply:
> This supposes that the Trait is used with a CommonObject and the Trait can not itself be restricted to certain classes language-wise.
> 
> There are two solutions:
> 
> * Test that $this is a CommonObject (`if ($this instanceof CommonObject`) and preferable add some error handling;
> * Add phpdoc suggesting that the method exists (which can also serve as documentation to remind that the method is expected on class the trait is used with).
> 
> ```
> /**
>  * Methods that should exist in the host class:
>  * @method call_trigger() <returntype> call_trigger(<param doc>) 
>  *
>  * To detect all properties that are supposed to exist in the host class:
>  * @phan-forbid-undeclared-magic-properties 
>  */
> ```
> 
> Regarding the properties, you can suggest them with `@property` in the class documentation.

Thanks for advice !